In [1]:
import pandas as pd

# Loading CSV

In [2]:
! ls /home/dsc/Data/us_dot/traffic

T100_SEGMENT_ALL_CARRIER_2014.zip  T100_SEGMENT_ALL_CARRIER_2015.zip


In [3]:
! unzip /home/dsc/Data/us_dot/traffic/T100_SEGMENT_ALL_CARRIER_2015.zip

Archive:  /home/dsc/Data/us_dot/traffic/T100_SEGMENT_ALL_CARRIER_2015.zip
  inflating: 914310910_T_T100_SEGMENT_ALL_CARRIER_2015_All.csv  
  inflating: 914310910_T_T100_SEGMENT_ALL_CARRIER_ReadMe.csv  
  inflating: 914310910_T_T100_SEGMENT_ALL_CARRIER_Terms.csv  


In [4]:
!pwd

/home/dsc/Repositories/masterdatascience/Python


In [6]:
df = pd.read_csv('914310910_T_T100_SEGMENT_ALL_CARRIER_2015_All.csv')
df.head()

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED  PAYLOAD  SEATS  PASSENGERS  \
0                   0.0                   1.0    528.0   11.0         2.0   
1                   0.0                   2.0   3621.0   22.0         2.0   
2                   0.0                   1.0   2169.0   11.0         2.0   
3                   0.0                   1.0   1342.0   11.0         1.0   
4                   0.0                   1.0    275.0   11.0         1.0   

   FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME     ...       \
0      0.0   0.0    3899.0           0.0       0.0     ...        
1      0.0   0.0    2589.0           0.0       0.0     ...        
2      0.0   0.0    1210.0           0.0       0.0     ...        
3      0.0   0.0    2316.0           0.0       0.0     ...        
4      0.0   0.0    1023.0           0.0       0.0     ...        

  AIRCRAFT_GROUP  AIRCRAFT_TYPE AIRCRAFT_CONFIG  YEAR QUARTER MONTH  \
0              6            669               1  2015       1     1   
1              6            669               1  2015       1     1   
2              6            669               1  2015       1     1   
3              6            669               1  2015       1     1   
4              6            669               1  2015       1     1   

  DISTANCE_GROUP  CLASS  DATA_SOURCE  Unnamed: 50  
0              8      L           IF          NaN  
1              6      L           IF          NaN  
2              3      L           DF          NaN  
3              5      L           IF          NaN  
4              3      L           DF          NaN  

[5 rows x 51 columns]

In [8]:
len(df)

146938

In [14]:
%%timeit
df[df['DEPARTURES_PERFORMED'] >50]['PAYLOAD'].sum()

5.72 ms ± 73.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# Saving to Excel

In [17]:
df[df['DEPARTURES_PERFORMED'] >50].to_excel('out.xlsx')

# Saving to CSV

In [18]:
df.to_csv('out.csv')

In [21]:
file_as_string = df.to_csv()
file_as_string[:100]

',DEPARTURES_SCHEDULED,DEPARTURES_PERFORMED,PAYLOAD,SEATS,PASSENGERS,FREIGHT,MAIL,DISTANCE,RAMP_TO_RA'

# To Sql Database

In [22]:
import sqlite3

conn = sqlite3.connect('example.db')

In [23]:
df.to_sql('traffic',conn)

/home/dsc/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:1362: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  chunksize=chunksize, dtype=dtype)


# To dictionary and to json

In [25]:
df[:3].to_dict()

{'AIRCRAFT_CONFIG': {0: 1, 1: 1, 2: 1},
 'AIRCRAFT_GROUP': {0: 6, 1: 6, 2: 6},
 'AIRCRAFT_TYPE': {0: 669, 1: 669, 2: 669},
 'AIRLINE_ID': {0: 21407.0, 1: 21407.0, 2: 21407.0},
 'AIR_TIME': {0: 0.0, 1: 0.0, 2: 0.0},
 'CARRIER': {0: '10Q', 1: '10Q', 2: '10Q'},
 'CARRIER_GROUP': {0: 0, 1: 0, 2: 0},
 'CARRIER_GROUP_NEW': {0: 0.0, 1: 0.0, 2: 0.0},
 'CARRIER_NAME': {0: 'Swiss Air Ambulance',
  1: 'Swiss Air Ambulance',
  2: 'Swiss Air Ambulance'},
 'CLASS': {0: 'L', 1: 'L', 2: 'L'},
 'DATA_SOURCE': {0: 'IF', 1: 'IF', 2: 'DF'},
 'DEPARTURES_PERFORMED': {0: 1.0, 1: 2.0, 2: 1.0},
 'DEPARTURES_SCHEDULED': {0: 0.0, 1: 0.0, 2: 0.0},
 'DEST': {0: 'BSL', 1: 'TEB', 2: 'MFE'},
 'DEST_AIRPORT_ID': {0: 10762, 1: 15167, 2: 13256},
 'DEST_AIRPORT_SEQ_ID': {0: 1076203, 1: 1516702, 2: 1325602},
 'DEST_CITY_MARKET_ID': {0: 30762, 1: 35167, 2: 33256},
 'DEST_CITY_NAME': {0: 'Mulhouse, France',
  1: 'Teterboro, NJ',
  2: 'Mission/McAllen/Edinburg, TX'},
 'DEST_COUNTRY': {0: 'FR', 1: 'US', 2: 'US'},
 'DEST_COUN

In [26]:
df[:3].to_dict(orient='rows')

[{'AIRCRAFT_CONFIG': 1,
  'AIRCRAFT_GROUP': 6,
  'AIRCRAFT_TYPE': 669,
  'AIRLINE_ID': 21407.0,
  'AIR_TIME': 0.0,
  'CARRIER': '10Q',
  'CARRIER_GROUP': 0,
  'CARRIER_GROUP_NEW': 0.0,
  'CARRIER_NAME': 'Swiss Air Ambulance',
  'CLASS': 'L',
  'DATA_SOURCE': 'IF',
  'DEPARTURES_PERFORMED': 1.0,
  'DEPARTURES_SCHEDULED': 0.0,
  'DEST': 'BSL',
  'DEST_AIRPORT_ID': 10762,
  'DEST_AIRPORT_SEQ_ID': 1076203,
  'DEST_CITY_MARKET_ID': 30762,
  'DEST_CITY_NAME': 'Mulhouse, France',
  'DEST_COUNTRY': 'FR',
  'DEST_COUNTRY_NAME': 'France',
  'DEST_STATE_ABR': nan,
  'DEST_STATE_FIPS': nan,
  'DEST_STATE_NM': nan,
  'DEST_WAC': 427,
  'DISTANCE': 3899.0,
  'DISTANCE_GROUP': 8,
  'FREIGHT': 0.0,
  'MAIL': 0.0,
  'MONTH': 1,
  'ORIGIN': 'EWR',
  'ORIGIN_AIRPORT_ID': 11618,
  'ORIGIN_AIRPORT_SEQ_ID': 1161802,
  'ORIGIN_CITY_MARKET_ID': 31703,
  'ORIGIN_CITY_NAME': 'Newark, NJ',
  'ORIGIN_COUNTRY': 'US',
  'ORIGIN_COUNTRY_NAME': 'United States',
  'ORIGIN_STATE_ABR': 'NJ',
  'ORIGIN_STATE_FIPS': 34.0,

In [27]:
df[:3].to_json()

'{"DEPARTURES_SCHEDULED":{"0":0.0,"1":0.0,"2":0.0},"DEPARTURES_PERFORMED":{"0":1.0,"1":2.0,"2":1.0},"PAYLOAD":{"0":528.0,"1":3621.0,"2":2169.0},"SEATS":{"0":11.0,"1":22.0,"2":11.0},"PASSENGERS":{"0":2.0,"1":2.0,"2":2.0},"FREIGHT":{"0":0.0,"1":0.0,"2":0.0},"MAIL":{"0":0.0,"1":0.0,"2":0.0},"DISTANCE":{"0":3899.0,"1":2589.0,"2":1210.0},"RAMP_TO_RAMP":{"0":0.0,"1":0.0,"2":0.0},"AIR_TIME":{"0":0.0,"1":0.0,"2":0.0},"UNIQUE_CARRIER":{"0":"10Q","1":"10Q","2":"10Q"},"AIRLINE_ID":{"0":21407.0,"1":21407.0,"2":21407.0},"UNIQUE_CARRIER_NAME":{"0":"Swiss Air Ambulance","1":"Swiss Air Ambulance","2":"Swiss Air Ambulance"},"UNIQUE_CARRIER_ENTITY":{"0":"71084","1":"71084","2":"71084"},"REGION":{"0":"I","1":"I","2":"I"},"CARRIER":{"0":"10Q","1":"10Q","2":"10Q"},"CARRIER_NAME":{"0":"Swiss Air Ambulance","1":"Swiss Air Ambulance","2":"Swiss Air Ambulance"},"CARRIER_GROUP":{"0":0,"1":0,"2":0},"CARRIER_GROUP_NEW":{"0":0.0,"1":0.0,"2":0.0},"ORIGIN_AIRPORT_ID":{"0":11618,"1":12651,"2":12889},"ORIGIN_AIRPORT_S

json is the format that javascript uses for storing objects

# Reading Excel

In [36]:
df2 = pd.read_excel('out.xls')
df2.head()

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED  PAYLOAD  SEATS  PASSENGERS  \
37                      0                    61  5365411  18353       16933   
42                      0                    61  5365411  18353       16827   
158                     0                    51  2706103  17384       13517   
159                     0                    51  2751949  17384       13746   
181                     0                    62  9292527  20584       16739   

     FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME     ...       \
37    571685     0      4395             0         0     ...        
42   1274568     0      4395             0         0     ...        
158        0     0      2002             0         0     ...        
159        0     0      2002             0         0     ...        
181  1505486     0      6504             0         0     ...        

    AIRCRAFT_GROUP  AIRCRAFT_TYPE AIRCRAFT_CONFIG  YEAR QUARTER MONTH  \
37               6            627               1  2015       1     1   
42               6            627               1  2015       1     1   
158              6            627               1  2015       1     1   
159              6            627               1  2015       1     1   
181              6            637               1  2015       1     1   

    DISTANCE_GROUP  CLASS  DATA_SOURCE  Unnamed: 50  
37               9      F           IF          NaN  
42               9      F           IF          NaN  
158              5      F           IF          NaN  
159              5      F           IF          NaN  
181             14      F           IF          NaN  

[5 rows x 51 columns]

In [45]:
path = '/home/dsc/Downloads/01_ ACCIDENTES POR TIPO EN  DISTRITOS.xls'
accidentes = pd.read_excel(path, index_col=0,skiprows = 7,skip_footer=1,parse_cols = 'A:K')
accidentes.head(10)

COLISIÓN DOBLE                            \
DISTRITO_ACCIDENTE                                                         
ARGANZUELA                                                           389   
BARAJAS                                                               89   
CARABANCHEL                                                          375   
CENTRO                                                               514   
CHAMARTIN                                                            494   
CHAMBERI                                                             414   
CIUDAD LINEAL                                                        398   
FUENCARRAL-EL PARDO                                                  321   
HORTALEZA                                                            215   
LATINA                                                               298   

                                COLISIÓN MÚLTIPLE                         \
DISTRITO_ACCIDENTE                                                         
ARGANZUELA                                                            54   
BARAJAS                                                                6   
CARABANCHEL                                                           44   
CENTRO                                                                55   
CHAMARTIN                                                             70   
CHAMBERI                                                              35   
CIUDAD LINEAL                                                         64   
FUENCARRAL-EL PARDO                                                   47   
HORTALEZA                                                             16   
LATINA                                                                40   

                                CHOQUE CON OBJETO FIJO                    \
DISTRITO_ACCIDENTE                                                         
ARGANZUELA                                                           126   
BARAJAS                                                               53   
CARABANCHEL                                                          171   
CENTRO                                                               171   
CHAMARTIN                                                            133   
CHAMBERI                                                              87   
CIUDAD LINEAL                                                        124   
FUENCARRAL-EL PARDO                                                  151   
HORTALEZA                                                            109   
LATINA                                                               134   

                                ATROPELLO                                 \
DISTRITO_ACCIDENTE                                                         
ARGANZUELA                                                            75   
BARAJAS                                                               21   
CARABANCHEL                                                          137   
CENTRO                                                               143   
CHAMARTIN                                                             92   
CHAMBERI                                                              87   
CIUDAD LINEAL                                                        101   
FUENCARRAL-EL PARDO                                                   78   
HORTALEZA                                                             58   
LATINA                                                                85   

                                VUELCO                                    \
DISTRITO_ACCIDENTE                                                         
ARGANZUELA                                                           9.0   
BARAJAS                                                              NaN   
CARABANCHEL                                                          8.0   


#### Exercise: Play with formats

In [3]:
import requests
multas_response = requests.get('http://datos.madrid.es/egob/catalogo/210104-170-multas-circulacion-detalle.csv')
multas_path = 'multas_2017_09.csv'
multas_file = open(multas_path,'wb')
multas_file.write(multas_response.content)
multas_file.close()

In [2]:
multas_path = 'multas_2017_09.csv'
df_multas = pd.read_csv(multas_path)

ParserError: Error tokenizing data. C error: Expected 3 fields in line 46859, saw 4


It is wrong made. There are four columns in line 46859

In [5]:
! wc -l {multas_path}

248411 multas_2017_09.csv


In [6]:
! head {multas_path}

The separator is ; , apart from the spaces, and the codification is UTF-8

In [3]:
df_multas = pd.read_csv(multas_path, sep=';',encoding='latin-1')
df_multas.head()

CALIFICACION                                     LUGAR  MES  ANIO   HORA  \
0   GRAVE       MARIA BARRIENTOS 4                          9  2017  19.15   
1   MUY GRAVE   AV MEMORIA FR 46                            9  2017   6.00   
2   LEVE        CL CARLOS LATORRE 15                        9  2017  10.56   
3   GRAVE       KM 23,9 M-30 CALZADA 2                      9  2017  18.43   
4   MUY GRAVE   AV MEMORIA FR 46                            9  2017   5.28   

   IMP_BOL DESCUENTO  PUNTOS           DENUNCIANTE  \
0    200.0        SI       0  POLICIA MUNICIPAL      
1    500.0        SI       4  POLICIA MUNICIPAL      
2     90.0        SI       0  SER                    
3    100.0        SI       0  POLICIA MUNICIPAL      
4    500.0        SI       4  POLICIA MUNICIPAL      

                                           HECHO-BOL VEL_LIMITE VEL_CIRCULA  \
0  ESTACIONAR SOBRE O JUNTO A REFUGIO, ISLETA, ME...                          
1  CONDUCIR CON UNA TASA DE ALCOHOLEMIA NO PERMIT...                          
2  ESTACIONAR, SIN LA CORRESPONDIENTE AUTORIZACIÓ...                          
3  SOBREPASAR LA VELOCIDADMÁXIMA EN VÍAS LIMITADA...         90         101   
4  CONDUCIR CON UNA TASA DE ALCOHOLEMIA NO PERMIT...                          

  COORDENADA_X COORDENADA_Y  
0                            
1                            
2                            
3                            
4

In [5]:
df_multas['LUGAR'].loc[4]

'AV MEMORIA FR 46                        '

In [7]:
df_multas[df_multas['LUGAR'] == 'AV MEMORIA FR 46                        ']

CALIFICACION                                     LUGAR  MES  ANIO  HORA  \
1   MUY GRAVE   AV MEMORIA FR 46                            9  2017  6.00   
4   MUY GRAVE   AV MEMORIA FR 46                            9  2017  5.28   

   IMP_BOL DESCUENTO  PUNTOS           DENUNCIANTE  \
1    500.0        SI       4  POLICIA MUNICIPAL      
4    500.0        SI       4  POLICIA MUNICIPAL      

                                           HECHO-BOL VEL_LIMITE VEL_CIRCULA  \
1  CONDUCIR CON UNA TASA DE ALCOHOLEMIA NO PERMIT...                          
4  CONDUCIR CON UNA TASA DE ALCOHOLEMIA NO PERMIT...                          

  COORDENADA_X COORDENADA_Y  
1                            
4

In [11]:
df_multas['IMP_BOL'].mean()

100.04955517088683

In [ ]:
df_multas['IMP_BOL'].sum()

#### Exercise

Calculate the amount that the ciity hall took in, after discount. Assume that the discount is 50%, for each line that has 'SI' in the column 'DESCUENTO'

In [7]:
df_multas['DESCUENTO_PCT'] = .5* (df_multas['DESCUENTO'] == 'SI')
df_multas['DESCUENTO_PCT'].head()


0    0.5
1    0.5
2    0.5
3    0.5
4    0.5
Name: DESCUENTO_PCT, dtype: float64

In [8]:
df_multas['IMPORTE_FINAL'] = df_multas['IMP_BOL'] * (1 - df_multas['DESCUENTO_PCT'])

#### Exercise

Consolidate the excel into one DataFrame: you will need to create a 'YEAR' column. Think how you can iterate through all the DataFrames. Think also where you can take the valueof 'YEAR' from.

For downloading something from internet:

In [42]:
import requests

response = requests.get('http://datos.madrid.es/egob/catalogo/207831-0-accidentes-trafico.xls')
out = open('accidentes.xls','wb')
out.write(response.content)
out.close

<function BufferedWriter.close>

# Reading mysql database

In [46]:
conn

In [48]:
pd.read_sql_query('SELECT * FROM traffic WHERE SEATS > 160', conn)

index  DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED     PAYLOAD  \
0          14                   0.0                   9.0    562962.0   
1          15                   0.0                  18.0   1237995.0   
2          16                   0.0                   9.0    565173.0   
3          17                   0.0                  18.0   1237436.0   
4          18                   0.0                   4.0    404800.0   
5          19                   0.0                   9.0    910800.0   
6          20                   0.0                   4.0    404800.0   
7          21                   0.0                   9.0    910800.0   
8          22                   0.0                  23.0   2345933.0   
9          23                   0.0                  29.0   2557427.0   
10         24                   0.0                  12.0   1049765.0   
11         25                   0.0                  23.0   2401058.0   
12         26                   0.0                  21.0   2199395.0   
13         27                   0.0                  29.0   3020211.0   
14         28                   0.0                  22.0   2294431.0   
15         29                   0.0                  14.0   1466137.0   
16         30                   0.0                  12.0   1250181.0   
17         31                   0.0                  21.0   2146910.0   
18         32                   0.0                  22.0   1924391.0   
19         33                   0.0                  14.0   1435498.0   
20         37                   0.0                  61.0   5365411.0   
21         40                   0.0                  27.0   2251555.0   
22         41                   0.0                   2.0    225601.0   
23         42                   0.0                  61.0   5365411.0   
24         43                   0.0                  28.0   2333124.0   
25         44                   0.0                   2.0    225601.0   
26         45                   0.0                   1.0     83875.0   
27         46                   0.0                   2.0    138283.0   
28         47                   0.0                  20.0   1677500.0   
29         48                   0.0                   5.0    407847.0   
...       ...                   ...                   ...         ...   
90109  146908                 522.0                 522.0  16314400.0   
90110  146909                 525.0                 518.0    932400.0   
90111  146910                 525.0                 524.0  16386600.0   
90112  146911                 526.0                 525.0  16411100.0   
90113  146912                 531.0                 531.0    637200.0   
90114  146913                 535.0                 534.0  16690700.0   
90115  146914                 538.0                 535.0  16727800.0   
90116  146915                 540.0                 538.0  16820600.0   
90117  146916                 540.0                 540.0  16886100.0   
90118  146917                 563.0                 563.0  17612200.0   
90119  146918                 565.0                 565.0  17672100.0   
90120  146919                 581.0                 580.0  18142300.0   
90121  146920                 586.0                 584.0  18261300.0   
90122  146921                 613.0                 603.0  10733400.0   
90123  146922                 613.0                 605.0  10769000.0   
90124  146923                 648.0                  23.0   4907970.0   
90125  146924                 668.0                 661.0  11765800.0   
90126  146925                 674.0                 664.0  11819200.0   
90127  146926                 732.0                 731.0  22832200.0   
90128  146927                 732.0                 728.0  22741700.0   
90129  146928                 741.0                 740.0  23118800.0   
90130  146929                 742.0                 741.0  23148900.0   
90131  146930                 743.0                 726.0  1292280